File: site/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/wallet/GetOfferingsTest.kt

package website.tbd.developer.site.docs.tbdex.wallet

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import tbdex.sdk.httpclient.TbdexHttpClient
import tbdex.sdk.protocol.models.Offering
import tbdex.sdk.protocol.serialization.Json
import java.net.HttpURLConnection
import website.tbd.developer.site.docs.utils.*

import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * Tests for Wallet: Get Offerings guide
 */
class GetOfferingsTest {

    private lateinit var server: MockWebServer
    private val pfi = TestData.PFI_DID
    private val pfiDid = pfi.uri

    @BeforeEach
    fun setup() {
        server = MockWebServer()
        server.start(9000) // pfiDid resolves to http://localhost:9000

        val offering = TestData.getOffering(pfiDid, TestData.getPresentationDefinition())
        offering.sign(pfi)
        val mockOfferings = listOf(offering)
        val mockResponseString = Json.jsonMapper.writeValueAsString(mapOf("data" to mockOfferings))
        server.enqueue(MockResponse().setBody(mockResponseString).setResponseCode(HttpURLConnection.HTTP_OK))
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `get all offerings`() {
        // :snippet-start: walletGetOfferingsKt
        val offerings =  TbdexHttpClient.getOfferings(pfiDid)
        // :snippet-end:

        assertNotEquals(0, offerings.size, "No offerings found from PFI")
    }

    @Test
    fun `find matching offerings`() {
        val pfiDids = arrayOf(pfiDid)

        // :snippet-start: walletFindMatchingOfferingsKt
        val payinCurrencyCode = "USD"; // Desired payin currency code
        val payoutCurrencyCode = "KES"; // Desired payout currency code

        val matchedOfferings = ArrayList<Offering>() // Array to store the matched offerings

        // Loop through the all PFIs in your network
        for (pfiDid in pfiDids) {

            // Makes a request to the PFI to get their offerings
            val offerings = TbdexHttpClient.getOfferings(pfiDid)

            // Filter offerings based on the currency pair
            val filteredOfferings = offerings.filter { offering ->
                offering.data.payin.currencyCode == payinCurrencyCode &&
                offering.data.payout.currencyCode == payoutCurrencyCode
            }

            matchedOfferings.addAll(filteredOfferings)
        }
        // :snippet-end:

        assertNotEquals(0, matchedOfferings.size, "No matching offerings found")
    }
}

File: site-new/testsuites/testsuite-javascript/__tests__/tbdex/wallet/getOfferings.test.js
import { vi, test, expect, describe, beforeAll, afterAll } from 'vitest';
import { TbdexHttpClient, DevTools } from '@tbdex/http-client';
import { DidDht } from '@web5/dids';
import { setupServer } from 'msw/node'
import { HttpResponse, http } from 'msw'

let pfi;
let pfiDid; //The URI of the PFI's DID
let server;
let mockOffering;

describe('Wallet: Get Offerings from PFI', () => {

  beforeAll(async () => {
    pfi = await DidDht.create({
      options:{
        publish: true,
        services: [{
            id: 'pfi',
            type: 'PFI',
            serviceEndpoint: 'https://localhost:9000'
        }]
      }
    })
    pfiDid = pfi.uri;

    // Mock the response from the PFI
    const defaultOfferingData = DevTools.createOfferingData()
    mockOffering = DevTools.createOffering({
      from: pfiDid,
      offeringData: {
          ...defaultOfferingData,
          payin: {
              ...defaultOfferingData.payin,
              currencyCode: 'USD'
          },
          payout: {
              ...defaultOfferingData.payout,
              currencyCode: 'KES'
          }
        }
    });
    await mockOffering.sign(pfi)

    server = setupServer(
      http.get('https://localhost:9000/offerings', () => {
        return HttpResponse.json({ data: [mockOffering] }, {
          status: 200
        })
      }),
    )
    server.listen({onUnhandledRequest: 'bypass'})
  });

  afterAll(() => {
    server.resetHandlers()
    server.close()
  });

  test('get all offerings', async () => {
    // :snippet-start: walletGetOfferingsJS
    const offerings  = await TbdexHttpClient.getOfferings({ pfiDid: pfiDid });
    // :snippet-end:
    expect(offerings).toEqual([mockOffering]);
  });

  test('find matching offerings', async () => {
    const pfiDids = [pfiDid];

    // :snippet-start: walletFindMatchingOfferingsJS
    const payinCurrencyCode = 'USD'; // Desired payin currency code
    const payoutCurrencyCode = 'KES'; // Desired payout currency code

    const matchedOfferings = []; // Array to store the matched offerings

    // Loop through the all PFIs in your network
    for (const pfiDid of pfiDids) {

      //Makes a request to the PFI to get their offerings
      const offerings = await TbdexHttpClient.getOfferings({ pfiDid: pfiDid });

      // Filter offerings based on the currency pair
      if(offerings){
        const filteredOfferings = offerings.filter(offering =>
          offering.data.payin.currencyCode === payinCurrencyCode &&
          offering.data.payout.currencyCode === payoutCurrencyCode
        );
        matchedOfferings.push(...filteredOfferings);
      }
    }
    // :snippet-end:

    expect(matchedOfferings).toEqual([mockOffering]);
  });
});